File: q3408/q3408.java
class Task {
    int taskId;
    int priority;

    public Task(int taskId, int priority) {
        this.taskId = taskId;
        this.priority = priority;
    }
}
class TaskManager {
    private TreeMap<Task, Integer> taskQueue;
    private Map<Integer, Integer> taskPriorityMap;

    public TaskManager(List<List<Integer>> tasks) {
        taskQueue = new TreeMap<Task, Integer>((a, b) -> {
            if (a.priority != b.priority) {
                return b.priority - a.priority; // higher priority first
            }
            return b.taskId - a.taskId;
        });

        taskPriorityMap = new HashMap<>();

        for (List<Integer> task : tasks) {
            int userId = task.get(0);
            int taskId = task.get(1);
            int priority = task.get(2);
            taskQueue.put(new Task(taskId, priority), userId);
            taskPriorityMap.put(taskId, priority);
        }
    }
    
    public void add(int userId, int taskId, int priority) {
        taskQueue.put(new Task(taskId, priority), userId);
        taskPriorityMap.put(taskId, priority);
    }
    
    public void edit(int taskId, int newPriority) {
        int oldPriority = taskPriorityMap.get(taskId);
        Task oldTask = new Task(taskId, oldPriority);
        int userId = taskQueue.get(oldTask);

        taskQueue.remove(oldTask);
        add(userId, taskId, newPriority);
    }
    
    public void rmv(int taskId) {
        int priority = taskPriorityMap.get(taskId);
        taskQueue.remove(new Task(taskId, priority));
        taskPriorityMap.remove(taskId);
    }
    
    public int execTop() {
        if (taskQueue.isEmpty()) return -1;
        Map.Entry<Task, Integer> topTaskEntry = taskQueue.firstEntry();
        rmv(topTaskEntry.getKey().taskId);
        return topTaskEntry.getValue();
    }
}

File: q3408/q3408.py
class Task:
    def __init__(self, userId, taskId, priority):
        self.userId = userId
        self.taskId = taskId
        self.priority = priority
        self.isDone = False
    
    def __lt__(self, other): 
        if self.priority == other.priority:
            return self.taskId > other.taskId  # or some consistent tie-breaker
        return self.priority > other.priority

class TaskManager:

    def __init__(self, tasks: List[List[int]]):

        self.tasksPQ = []
        self.taskIdToTaskMap = dict()

        for t in tasks: 
            task = Task(t[0], t[1], t[2])
            self.taskIdToTaskMap[t[1]] = task 
            heappush(self.tasksPQ, task)
        
    def add(self, userId: int, taskId: int, priority: int) -> None:
        task = Task(userId, taskId, priority)
        self.taskIdToTaskMap[taskId] = task
        heappush(self.tasksPQ, task)

    def edit(self, taskId: int, newPriority: int) -> None:
        oldTask = self.taskIdToTaskMap[taskId]
        oldTask.isDone = True  # Mark old task as done

        # Create and insert new task with updated priority
        newTask = Task(oldTask.userId, taskId, newPriority)
        self.taskIdToTaskMap[taskId] = newTask
        heappush(self.tasksPQ, newTask)

    def rmv(self, taskId: int) -> None:
        task = self.taskIdToTaskMap[taskId]
        task.isDone = True

    def execTop(self) -> int:
        while self.tasksPQ:
            top = heappop(self.tasksPQ)
            if not top.isDone:
                top.isDone = True
                del self.taskIdToTaskMap[top.taskId]
                return top.userId
        return -1

# Your TaskManager object will be instantiated and called as such:
# obj = TaskManager(tasks)
# obj.add(userId,taskId,priority)
# obj.edit(taskId,newPriority)
# obj.rmv(taskId)
# param_4 = obj.execTop()